### ISSUE TITLE
7$: Report error on BOM before #!

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#205]
BOM: Byte Order Mark
#!: Shebang

Linux (and I assume macOS and other POSIX-like platforms) only interpret shebangs if the first byte in the file is `#` and the second byte in the file is `!` (I think; please confirm). This means that the presence of a BOM (UTF-8-encoded) will cause the shebang characters to not be interpreted by the operating system. This is almost definitely a mistake.

Report an error if this mistake occurs.

**Comment by singalhimanshu** [comment]
I claim this for-hire task. I expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

**Comment by singalhimanshu** [comment]
Is this the correct place to look in for this issue? - https://github.com/quick-lint/quick-lint-js/blob/8ab6bed96f16678b686395643728e580e6e8d0e0/src/lex.cpp#L543

Right now for this program (with BOM character before #) - 
```javascript
#!/usr/bin/env node
let x = 3;
console.log(x);
```
quick-lint-js report the following errors and warnings - 
```shell
main.js:1:4: error: unexpected '#' [E052]
main.js:1:18: error: missing semicolon after statement [E027]
main.js:1:15: warning: use of undeclared variable: env [E057]
main.js:1:19: warning: use of undeclared variable: node [E057]
```
What error should is expected to be reported?

**Comment by strager** [comment]
> Is this the correct place to look in for this issue? -
> 
> https://github.com/quick-lint/quick-lint-js/blob/8ab6bed96f16678b686395643728e580e6e8d0e0/src/lex.cpp#L543

I don't think so. When we execute that piece of code, we skipped the BOM already.

> Right now for this program (with BOM character before #) -
> 
> ```js
> #!/usr/bin/env node
> let x = 3;
> console.log(x);
> ```
> 
> quick-lint-js report the following errors and warnings -
> 
> ```shell
> main.js:1:4: error: unexpected '#' [E052]
> main.js:1:18: error: missing semicolon after statement [E027]
> main.js:1:15: warning: use of undeclared variable: env [E057]
> main.js:1:19: warning: use of undeclared variable: node [E057]
> ```
> 
> What error should is expected to be reported?

`error: Unicode byte order mark (BOM) cannot appear before #! at beginning of script`
